codeworld-server/src/Build.hs

```diff
@@ -29,118 +31,28 @@ import           System.IO.Temp (withSystemTempDirectory)
 import           System.Process
 import           Text.Regex.TDFA
 
-import ErrorSanitizer
 import Util
 
 compileIfNeeded :: BuildMode -> ProgramId -> IO Bool
 compileIfNeeded mode programId = do
     hasResult <- doesFileExist (buildRootDir mode </> resultFile programId)
     hasTarget <- doesFileExist (buildRootDir mode </> targetFile programId)
-    if hasResult then return hasTarget else compileExistingSource mode programId
+    if hasResult 
+        then return hasTarget 
+        else C.compileSource 
+                 (sourceAddr mode programId) 
+                 (compiledAddr mode programId) 
+                 (errorAddr mode programId) 
+                 (getMode mode) 
 
-compileExistingSource :: BuildMode -> ProgramId -> IO Bool
-compileExistingSource mode programId = checkDangerousSource mode programId >>= \case
-    True -> do
-        B.writeFile (buildRootDir mode </> resultFile programId) $
-            "Sorry, but your program refers to forbidden language features."
-        return False
-    False -> withSystemTempDirectory "codeworld" $ \tmpdir -> do
-        copyFile (buildRootDir mode </> sourceFile programId) (tmpdir </> "program.hs")
-        let baseArgs = case mode of
-                BuildMode "haskell"   -> haskellCompatibleBuildArgs
-                _                     -> standardBuildArgs
-            ghcjsArgs = baseArgs ++ [ "program.hs" ]
-        success <- runCompiler tmpdir userCompileMicros ghcjsArgs >>= \case
-            Nothing -> return False
-            Just output -> do
-                let filteredOutput = case mode of 
-                        BuildMode "haskell"   -> output
-                        _                     -> filterOutput output
-                B.writeFile (buildRootDir mode </> resultFile programId) filteredOutput
-                let target = tmpdir </> "program.jsexe" </> "all.js"
-                hasTarget <- doesFileExist target
-                when hasTarget $
-                    copyFile target (buildRootDir mode </> targetFile programId)
-                return hasTarget
-        return success
+sourceAddr :: BuildMode -> ProgramId -> FilePath
+sourceAddr mode programId = (buildRootDir mode </> sourceFile programId)
```

Comment:
Hmm.  I personally find the original expressions clearer than these wrappers, and not much longer.